internal/mempool/mempool.go

```diff
@@ -872,19 +876,183 @@ func (g *txDescGraph) forEachDescendant(txHash *chainhash.Hash,
+		if !moveNext {
+			return
+		}
+
+		seen[parentHash] = parentTxDesc
+		g.forEachAncestorPostOrder(&parentHash, seen, f)
+	}
+}
+
+// addAncestor adds the provided ancestor transaction's statistics
+// to the provided ancestor stats reference.
+func addAncestor(stats *mining.TxAncestorStats, ancestor *mining.TxDesc) {
+	stats.Fees += ancestor.Fee
+	stats.SizeBytes += ancestor.TxSize
+	stats.TotalSigOps += ancestor.TotalSigOps
+	stats.NumAncestors++
+}
+
+// maybeUpdateAncestorStats attempts to update the ancestor stats for a given
+// transaction. It walks the graph for all ancestors of the provided transaction
+// and aggregates statistics for its ancestors.
+func (mv *txMiningView) maybeUpdateAncestorStats(txDesc *mining.TxDesc) bool {
+	baseTxHash := txDesc.Tx.Hash()
+	canTrackAncestors := true
+	stats := &mining.TxAncestorStats{}
+	seenAncestors := make(map[chainhash.Hash]*mining.TxDesc,
+		ancestorTrackingLimit)
+	mv.txGraph.forEachAncestorPostOrder(baseTxHash, seenAncestors,
+		func(ancestorTxDesc *mining.TxDesc) bool {
+			if !canTrackAncestors {
+				// Short circuit the below checks if we know that the provided
+				// transaction cannot have ancestor stats aggregated. This also
+				// stops adding transactions to the seen map.
+				return false
+			}
+
+			if len(seenAncestors) >= ancestorTrackingLimit {
+				// A transaction with too many ancestors should not have
+				// ancestor stats cached.
+				canTrackAncestors = false
+				return false
+			}
+
+			// If any ancestor transaction does not have stats, then the
+			// provided one should not either.
+			ancestorTxHash := *ancestorTxDesc.Tx.Hash()
+			ancestorStats, exists := mv.ancestorStats[ancestorTxHash]
+			if !exists {
+				canTrackAncestors = false
+				return false
+			}
+
+			// If any ancestor plus its respective ancestors would exceed the
+			// limit for this transaction, then do not track stats.
+			if ancestorStats.NumAncestors+1 > ancestorTrackingLimit {
+				canTrackAncestors = false
+				return false
+			}
+
+			// If any ancestor has more children than the tracking limit,
+			// then stats should not be tracked for the provided transaction.
+			ancestorChildren := mv.txGraph.childrenOf[ancestorTxHash]
+			if len(ancestorChildren) >= ancestorTrackingLimit {
+				canTrackAncestors = false
+				return false
+			}
+
+			addAncestor(stats, ancestorTxDesc)
+			return true
+		})
+
+	if !canTrackAncestors {
+		delete(mv.ancestorStats, *baseTxHash)
+		return false
+	}
+
+	// The transaction has passed all checks and will have ancestor statistics
+	// tracked and updated. All elements of the seenAncestors map  will have
+	// ancestor statistics tracked. This is guaranteed by a check performed
+	// during a prior walk that limits what entered the seenAncestors map.
+	baseTxnAncestorStats := &mining.TxAncestorStats{}
+	mv.ancestorStats[*baseTxHash] = baseTxnAncestorStats
+	for _, ancestorTxDesc := range seenAncestors {
+		addAncestor(baseTxnAncestorStats, ancestorTxDesc)
+	}
```

Comment:
Hasn't this already all been collected in the `stats` variable above?

Reply:
Yes, `stats` should have been removed in a prior commit. Updated.